client/cache.go

```diff
@@ -0,0 +1,117 @@
+package client
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/log"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+// NewCachingClient is a meta client that stores an LRU cache of
+// recently fetched random values.
+func NewCachingClient(client Client, size int, log log.Logger) (Client, error) {
+	cache, err := lru.NewARC(size)
+	if err != nil {
+		return nil, err
+	}
+	return &cachingClient{
+		backing:     client,
+		cache:       cache,
+		log:         log,
+		subscribers: make([]subscriber, 0),
+	}, nil
+}
+
+type cachingClient struct {
+	backing Client
+	cache   *lru.ARCCache
+	log     log.Logger
+
+	subscriberLock sync.Mutex
+	subscribers    []subscriber
+}
+
+type subscriber struct {
+	ctx context.Context
+	c   chan Result
+}
+
+// Get returns a the randomness at `round` or an error.
+func (c *cachingClient) Get(ctx context.Context, round uint64) (res Result, err error) {
+	if val, ok := c.cache.Get(round); ok {
+		return val.(Result), nil
+	}
+	val, err := c.backing.Get(ctx, round)
+	if err == nil && val != nil {
+		c.cache.Add(round, val)
+	}
+	return val, err
+}
+
+// Watch returns new randomness as it becomes available.
+func (c *cachingClient) Watch(ctx context.Context) <-chan Result {
+	c.subscriberLock.Lock()
+	defer c.subscriberLock.Unlock()
+
+	sub := subscriber{ctx, make(chan Result, 5)}
+	c.subscribers = append(c.subscribers, sub)
+
+	if len(c.subscribers) == 1 {
+		ctx, cancel := context.WithCancel(context.Background())
+		go c.distribute(c.backing.Watch(ctx), cancel)
+	}
+	return sub.c
+}
+
+func (c *cachingClient) distribute(in <-chan Result, cancel context.CancelFunc) {
+	defer cancel()
+	for {
+		var aCtx context.Context
+		c.subscriberLock.Lock()
+		if len(c.subscribers) == 0 {
+			c.subscriberLock.Unlock()
+			return
+		}
+		aCtx = c.subscribers[0].ctx
+		c.subscriberLock.Unlock()
+
+		var m Result
+		select {
+		case res, ok := <-in:
+			if !ok {
+				return
+			}
+			c.cache.Add(res.Round(), res)
+			m = res
+		case <-aCtx.Done():
+		}
+
+		c.subscriberLock.Lock()
+		curr := c.subscribers
+		c.subscribers = c.subscribers[:0]
+
+		for _, s := range curr {
+			if s.ctx.Err() == nil {
+				c.subscribers = append(c.subscribers, s)
+				if m != nil {
+					select {
+					case s.c <- m:
+					default:
+						c.log.Warn("msg", "dropped watch message to subscriber. full channel")
```

Comment:
Is there a way to do this without dropping results?

Reply:
This only occurs when the client who has been given the channel isn't draining it. the channel can allow 5 results to back up internally (we could make that more), but that means nobody has pulled values off the channel for 2.5 minutes, which seems like a reasonable bound.